---
title: "Overlap of Differentially Expressed or Methylated Genes"
output: pdf_document
---
Summary report

```{r setup, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE}
require(knitr)
require(xtable)
opts_chunk$set(echo = FALSE, message=FALSE, warning=FALSE, cache=FALSE, results='markup')
source(file = "../src/de_genes_overlap.R")
writingDir <- "results/diffExpAnalysis/20150209/combined_mol_dat_types/"
```

## Target Phenotype: Preterm 
Compares differential express
### All Samples 
```{r loadDiffExpGeneLists}

targetPhenotypeName <- "Preterm"

# Inputs must be character vectors/factors
deGenesRNA <- read.table("results/diffExpAnalysis//20150209//combined_mol_dat_types/DE_RNASeq_Preterm_all_20150209.txt")
deGenesRNA <- deGenesRNA[,1]

deGenesMiRNA <- read.table("../test-data/differential-expression-genelist/targets-diffExp-mirs.txt")
deGenesMiRNA <- deGenesMiRNA[,1]

deGenesMeth <- read.table("../test-data/differential-expression-genelist/diff-meth-genelist.txt")
deGenesMeth <- deGenesMeth[,1]
  
pathwayList <- read.table("../test-data/differential-expression-genelist/genelist-pathways.txt", header=T, sep="\t")
geneList <- pathwayList$Gene
```

```{r overlapAnalysis, echo=FALSE, results='asis'}
# Returns summary of gene overlaps
res <- xtable(get_de_genes_overlap(deGenesRNA, deGenesMiRNA, deGenesMeth, geneList, targetPhenotypeName, writingDir))
print(res,  include.rownames=FALSE, type="latex", comment=FALSE)
```